trin-core/src/portalnet/find/iterators/findcontent.rs

```diff
@@ -0,0 +1,715 @@
+                    let peer = entry.get_mut();
+                    peer.set_state(QueryPeerState::Succeeded);
+                }
+                QueryPeerState::NotContacted
+                | QueryPeerState::Failed
+                | QueryPeerState::Succeeded => return,
+            },
+        }
+
+        // Incorporate the peer response into the query.
+        match peer_response {
+            FindContentResponse::ClosestNodes(closer_peers) => {
+                // Incorporate the reported closer peers into the query.
+                let mut progress = false;
+                let num_closest = self.closest_peers.len();
+
+                for peer in closer_peers {
+                    let key: Key<TNodeId> = peer.into();
+                    let distance = self.target_key.distance(&key);
+                    let peer = QueryPeer::new(key, QueryPeerState::NotContacted);
+                    self.closest_peers.entry(distance).or_insert(peer);
+
+                    // The query makes progress if the new peer is either closer to the target
+                    // than any peer seen so far (i.e. is the first entry), or the query did
+                    // not yet accumulate enough closest peers.
+                    progress = self.closest_peers.keys().next() == Some(&distance)
+                        || num_closest < self.config.num_results;
+                }
+
+                self.progress = match self.progress {
+                    QueryProgress::Iterating { no_progress } => {
+                        let no_progress = if progress { 0 } else { no_progress + 1 };
+                        if no_progress >= self.config.parallelism {
+                            QueryProgress::Stalled
+                        } else {
+                            QueryProgress::Iterating { no_progress }
+                        }
+                    }
+                    QueryProgress::Stalled => {
+                        if progress {
+                            QueryProgress::Iterating { no_progress: 0 }
+                        } else {
+                            QueryProgress::Stalled
+                        }
+                    }
+                    QueryProgress::Finished => QueryProgress::Finished,
+                };
+            }
+            FindContentResponse::Content(content) => {
+                // Incorporate the found content into the query.
+                self.content = Some(ContentAndPeer {
+                    content,
+                    peer: peer.clone(),
+                });
+            }
+        }
+    }
+
+    fn on_failure(&mut self, peer: &TNodeId) {
+        if let QueryProgress::Finished = self.progress {
+            return;
+        }
+
+        let key: Key<TNodeId> = peer.clone().into();
+        let distance = key.distance(&self.target_key);
+
+        match self.closest_peers.entry(distance) {
+            Entry::Vacant(_) => {}
+            Entry::Occupied(mut e) => match e.get().state() {
```

Comment:
*nit* I'm not a fan of single letter variables - imo `entry` would be better here.